✨ Initial support for conditions for AzureMachinePool

nprokopic opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind other
-->

**What this PR does / why we need it**:

Add `Status.Conditions` to `AzureMachinePool`.

`Status.Conditions` were added to `AzureCluster` and `AzureMachine` in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/714. With this change the resolution of issue #671, which also mentions `Status.Conditions` for `AzureMachinePool`, is more complete.

**Which issue(s) this PR fixes**:
Fixes partially #671 

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add `Status.Conditions` field to `AzureMachinePool` showing details about the current state of the object.
```

nprokopic commented:
/assign @justinsb

devigned commented:
/ok-to-test

devigned commented on exp/api/v1alpha3/azuremachinepool_types.go:
```diff
@@ -165,6 +170,16 @@ type (
 	}
 )
 
+// GetConditions returns the list of conditions for an AzureCluster API object.
```
```suggestion
// GetConditions returns the list of conditions for an AzureMachinePool API object.
```

devigned commented on exp/api/v1alpha3/azuremachinepool_types.go:
```diff
@@ -165,6 +170,16 @@ type (
+	return amp.Status.Conditions
+}
+
+// SetConditions will set the given conditions on an AzureCluster object
```
```suggestion
// SetConditions will set the given conditions on an AzureMachinePool object
```

devigned commented on config/crd/bases/exp.infrastructure.cluster.x-k8s.io_azuremachinepools.yaml:
```diff
@@ -296,8 +296,52 @@ spec:
             description: AzureMachinePoolStatus defines the observed state of AzureMachinePool
             properties:
+              conditions:
+                description: Conditions defines current service state of the AzureMachine.
```
```suggestion
                description: Conditions defines current service state of the AzureMachinePool.
```

devigned commented on exp/api/v1alpha3/azuremachinepool_types.go:
```diff
@@ -133,6 +134,10 @@ type (
 		// +optional
 		FailureMessage *string `json:"failureMessage,omitempty"`
+
+		// Conditions defines current service state of the AzureMachine.
```
```suggestion
		// Conditions defines current service state of the AzureMachinePool.
```

devigned reviewed (changes requested):
Looks like a bit of copypasta still left over in the comments.

nprokopic commented:
> Looks like a bit of copypasta still left over in the comments.

Well that's embarrassing :facepalm: Thank you for catching this, fixed it.

nprokopic commented:
This test failure:

```
sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3
  Incompatible changes:
  - AzureMachinePoolStatus: old is comparable, new is not
  Compatible changes:
  - (*AzureMachinePool).GetConditions: added
  - (*AzureMachinePool).SetConditions: added
  - AzureMachinePoolStatus.Conditions: added
```

Is this maybe due to `Conditions` (added to `AzureMachinePoolStatus`) being a slice?

Any advice here?

CecileRobertMichon commented:
@nprokopic the api diff job helps reviewers catch breaking changes to the API but it's not perfect. In this case, you're change affects the Status field of an experimental type so it's okay to ignore. The job is marked as optional so it won't be blocking to merge the PR.

nprokopic commented:
> @nprokopic the api diff job helps reviewers catch breaking changes to the API but it's not perfect. In this case, you're change affects the Status field of an experimental type so it's okay to ignore.

Thanks for the explanation!

> The job is marked as optional so it won't be blocking to merge the PR.

Great :)

CecileRobertMichon commented on exp/api/v1alpha3/azuremachinepool_types.go:
```diff
@@ -165,6 +170,16 @@ type (
+}
+
+// SetConditions will set the given conditions on an AzureMachinePool object
+func (amp *AzureMachinePool) SetConditions(conditions clusterv1.Conditions) {
```
Should this change also include defining basic conditions for AzureMachinePools and setting them in the controller like https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/714/files#diff-825ec7ff34f98855f2347c7f7769438dR284 did?

nprokopic replied:
Sounds reasonable. I'll look into it.

CecileRobertMichon replied:
@nprokopic are you still planning to do this? If not, we can merge this as-is and mark the issue as incomplete

nprokopic replied:
Yes, I do, just couldn't find the time in the previous weeks.

Can we merge this and have a separate pull request for setting conditions?

Lately I was working on conditions implementation for Giant Swarm's azure-operator (where we use CAPI and CAPZ types), here's what we added for current alpha release (so not fully completed yet) for Cluster, AzureCluster and MachinePool https://github.com/giantswarm/azure-operator/pull/1112, and here is current WIP for AzureMachinePool https://github.com/giantswarm/azure-operator/pull/1123.
Once we wrap these up, I would take a look at what's currently done in CAPZ, and then open an issue with suggestions for conditions that would make sense for CAPZ.

CecileRobertMichon replied:
Okay let's do that.

CecileRobertMichon reviewed (approved):
/lgtm
/approve
